lib/msal-core/src/ScopeSet.ts

```diff
@@ -1,69 +1,189 @@
-        return scopes.filter(value => value !== scope);
+     * @ignore
+     * Appends extraScopesToConsent if passed
+     * @param {@link AuthenticationParameters}
+     */
+    appendExtraScope(newScope: string): void {
+        this.scopes.add(newScope);
+    }
+
+    /**
+     * @ignore
+     * Appends extraScopesToConsent if passed
+     * @param {@link AuthenticationParameters}
+     */
+    appendExtraScopes(newScopes: Array<string>): void {
+        this.validateInputScopes(newScopes);
+        this.scopes = this.unionScopeSets(newScopes);
+    }
+
+    /**
+     * remove one element from set of scopes
+     *
+     * @param scope
+     */
+    removeElement(scope: string): void {
+        this.scopes.delete(scope);
+    }
+
+    /**
+     * @hidden
+     * @ignore
+     *
+     * ScopeSet helper to 
+     * @param matchingAccount boolean telling whether account object matches value in cache
+     * @param silentCall boolean telling whether or not silent call
+     *
+     * @returns {string} token type: id_token or access_token
+     *
+     */
+    getTokenType(matchingAccount: boolean, silentCall: boolean): string {
+        // if account is passed and matches the account object/or set to getAccount() from cache
+        // if client-id is passed as scope, get id_token else token/id_token_token (in case no session exists)
+        if (!matchingAccount) {
+            return silentCall && this.isLoginCall() ? ResponseTypes.id_token : ResponseTypes.id_token_token;           
+        } else {
+            return this.isLoginCall ? ResponseTypes.id_token : ResponseTypes.token;
+        }
+    }
+
+    /**
+     * ScopeSet helper which will tell whether set of scopes is being used for login calls.
+     */
+    isLoginCall(): boolean {
+        return this.originalScopes && this.originalScopes.has(this.clientId) && this.originalScopes.size === 1
+    }
+
+    /**
+     * @hidden
+     *
+     * Extracts scope value from the state sent with the authentication request.
+     * 
+     * Note: This function was removed from the current implementation of MSAL.js.
+     * @param {string} state
+     * @returns {string} scope.
+     * @ignore
+     */
+    // private getScopeFromState(state: string): string;
+
+    /**
+     * Returns the scopes as an array of string values.
+     */
+    asArray(): Array<string> {
+        return Array.from(this.scopes);
     }
 
     /**
-   * Parse the scopes into a formatted scopeList
-   * @param scopes
-   */
-    static parseScope(scopes: Array<string>): string {
+     * Print the scopes into a space-delimited string.
+     */
+    printScopes(): string {
         let scopeList: string = "";
-        if (scopes) {
-            for (let i: number = 0; i < scopes.length; ++i) {
-                scopeList += (i !== scopes.length - 1) ? scopes[i] + " " : scopes[i];
+        if (this.scopes) {
+            for (const scope in this.scopes) {
+                scopeList += scope + " ";
```

Comment:
I think it would cleaner to get the array version of the set, and then do `.join(" ")`